codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -572,46 +573,77 @@ private class ServerHttpProtocolImplGenerator(
         if (pathBindings.isEmpty()) {
             return
         }
-        val pattern = StringBuilder()
         val httpTrait = httpBindingResolver.httpTrait(operationShape)
-        httpTrait.uri.segments.forEach {
-            pattern.append("/")
-            if (it.isLabel) {
-                pattern.append("(?P<${it.content}>")
-                if (it.isGreedyLabel) {
-                    pattern.append(".+")
-                } else {
-                    pattern.append("[^/]+")
+        with(writer) {
+            val greedyLabelIndex = httpTrait.uri.segments.indexOfFirst { it.isGreedyLabel }
+            val segments =
+                if (greedyLabelIndex >= 0)
+                    httpTrait.uri.segments.slice(0 until (greedyLabelIndex + 1))
+                else
+                    httpTrait.uri.segments
+            val restAfterGreedyLabel =
+                if (greedyLabelIndex >= 0)
+                    httpTrait.uri.segments.slice((greedyLabelIndex + 1) until httpTrait.uri.segments.size).joinToString(prefix = "/", separator = "/")
+                else
+                    ""
+            val labeledNames = segments
+                .mapIndexed { index, segment ->
+                    if (segment.isLabel) { "m$index" } else { "_" }
                 }
-                pattern.append(")")
-            } else {
-                pattern.append(it.content)
+                .joinToString(separator = ",")
+            val nomParser = segments
+                .map { segment ->
+                    if (segment.isGreedyLabel) {
+                        "#{Nom}::combinator::rest"
+                    } else if (segment.isLabel) {
+                        """#{Nom}::branch::alt((#{Nom}::bytes::complete::take_until1("/"), #{Nom}::combinator::rest))"""
+                    } else {
+                        """#{Nom}::bytes::complete::tag("${segment.content}")"""
+                    }
+                }
+                .joinToString(
+                    prefix = "#{Nom}::sequence::tuple::<_, _, #{Nom}::error::Error<&str>, _>((",
```

Comment:
CI is failing here https://github.com/awslabs/smithy-rs/runs/4692069662?check_suite_focus=true#step:7:1064

In the case where there is only one label in the URI pattern and nothing else, using `sequence::tuple` doesn't seem to compile.